File: src-tauri/src/bin/lh.rs
fn main() {
    let args: Vec<String> = std::env::args().collect();
    let cmd = args.get(1).map(|s| s.as_str()).unwrap_or("list");

    match cmd {
        "list" => list_ports(),
        "check" => {
            let port = args.get(2).and_then(|s| s.parse::<u16>().ok()).unwrap_or(0);
            if port == 0 {
                eprintln!("Usage: lh check <port>");
                std::process::exit(1);
            }
            check_port(port);
        }
        "suggest" => {
            let start = args.get(2).and_then(|s| s.parse::<u16>().ok()).unwrap_or(3000);
            let end = args.get(3).and_then(|s| s.parse::<u16>().ok()).unwrap_or(3999);
            suggest_port(start, end);
        }
        "portmasters" => list_portmasters(),
        "projects" => list_project_ports(),
        _ => {
            eprintln!("Usage: lh [list|check <port>|suggest <start> <end>|portmasters|projects]");
            std::process::exit(1);
        }
    }
}

fn list_ports() {
    match lighthouse::scanner::scan_live_ports() {
        Ok(mut ports) => {
            println!("PORT   BIND            PROCESS");
            println!("-----  --------------  --------------------");
            ports.sort_by_key(|p| p.port);
            for p in ports {
                println!("{:<5}  {:<14}  {}", p.port, p.bind_address, p.process_name);
            }
        }
        Err(err) => {
            eprintln!("Failed to scan ports: {}", err);
            std::process::exit(1);
        }
    }
}

fn check_port(port: u16) {
    match lighthouse::scanner::check_single_port(port) {
        Ok(result) => {
            if result.in_use {
                println!("Port {} is in use", result.port);
                if let Some(proc_name) = result.process {
                    println!("Process: {}", proc_name);
                }
                if let Some(pid) = result.pid {
                    println!("PID: {}", pid);
                }
                if let Some(suggestion) = result.suggestion {
                    println!("Suggested free port: {}", suggestion);
                }
            } else {
                println!("Port {} is free", result.port);
            }
        }
        Err(err) => {
            eprintln!("Failed to check port {}: {}", port, err);
            std::process::exit(1);
        }
    }
}

fn suggest_port(start: u16, end: u16) {
    match lighthouse::resolver::suggest_port(start, end) {
        Ok(port) => println!("{}", port),
        Err(err) => {
            eprintln!("{}", err);
            std::process::exit(1);
        }
    }
}

fn list_portmasters() {
    match lighthouse::portmaster::discover_portmaster_files() {
        Ok(files) => {
            if files.is_empty() {
                println!("No PORTMASTER.md files found");
                return;
            }
            for file in files {
                println!("{}", file);
            }
        }
        Err(err) => {
            eprintln!("Failed to discover PORTMASTER.md files: {}", err);
            std::process::exit(1);
        }
    }
}

fn list_project_ports() {
    match lighthouse::projects::scan_project_ports() {
        Ok(mut ports) => {
            if ports.is_empty() {
                println!("No project port configs found");
                return;
            }
            ports.sort_by(|a, b| a.port.cmp(&b.port).then(a.project_name.cmp(&b.project_name)));
            println!("PORT   PROJECT                FILE");
            println!("-----  ---------------------  --------------------");
            for p in ports {
                println!("{:<5}  {:<21}  {}:{}", p.port, p.project_name, p.file_path, p.line_number);
            }
        }
        Err(err) => {
            eprintln!("Failed to scan project ports: {}", err);
            std::process::exit(1);
        }
    }
}

File: bin/lh.js
#!/usr/bin/env node

const { Command } = require('commander');
const { scanPorts, checkPort, suggestPort, getPortmasterFiles, friendlyProcessName } = require('../lib/engine');
const pkg = require('../package.json');

const program = new Command();

program
  .name('lh')
  .description('Lighthouse — local development port awareness')
  .version(pkg.version);

program
  .command('list')
  .description('list live ports')
  .option('--json', 'output as JSON')
  .action((opts) => {
    const { ports, conflicts } = scanPorts();
    if (opts.json) {
      console.log(JSON.stringify({ ports, conflicts }, null, 2));
      return;
    }
    if (ports.length === 0) {
      console.log('No listening ports found.');
      return;
    }
    console.log(`PORT   BIND          PROCESS              SERVICE            SOURCE`);
    console.log(`----   ----          -------              -------            ------`);
    for (const p of ports) {
      const port = String(p.port).padEnd(6);
      const bind = (p.bind_address || '').padEnd(13);
      const proc = friendlyProcessName(p.process_name).padEnd(20);
      const svc = (p.service_name || (p.undeclared ? '(undeclared)' : '')).padEnd(18);
      const flags = [];
      if (p.exposed) flags.push('exposed');
      if (p.undeclared) flags.push('undeclared');
      const src = (p.source || 'Live') + (flags.length ? ` [${flags.join(', ')}]` : '');
      console.log(`${port}${bind}${proc}${svc}${src}`);
    }
    if (conflicts.length > 0) {
      console.log(`\n⚠ ${conflicts.length} conflict(s) detected:`);
      for (const c of conflicts) {
        console.log(`  Port ${c.port}: ${c.explanation}`);
      }
    }
  });

program
  .command('check <port>')
  .description('check if a port is free')
  .option('--json', 'output as JSON')
  .action((portStr, opts) => {
    const port = parseInt(portStr, 10);
    if (isNaN(port) || port < 1 || port > 65535) {
      console.error(`Invalid port: ${portStr}`);
      process.exit(1);
    }
    const result = checkPort(port);
    if (opts.json) {
      console.log(JSON.stringify(result, null, 2));
      return;
    }
    if (result.in_use) {
      console.log(`Port ${port} is IN USE by ${friendlyProcessName(result.process)} (PID ${result.pid}) on ${result.bind}`);
    } else if (result.transient_state) {
      console.log(`Port ${port} is not listening but has a lingering ${result.transient_state} socket (${result.process}, PID ${result.pid})`);
    } else {
      console.log(`Port ${port} is free`);
    }
    if (result.suggestion) {
      console.log(`Suggested alternative: ${result.suggestion}`);
    }
  });

program
  .command('suggest')
  .description('suggest a free port')
  .option('-r, --range <start:end>', 'port range to search (default: 3000:3999)', '3000:3999')
  .option('--json', 'output as JSON')
  .action((opts) => {
    const [startStr, endStr] = opts.range.split(':');
    const start = parseInt(startStr, 10) || 3000;
    const end = parseInt(endStr, 10) || 3999;
    const port = suggestPort(start, end);
    if (opts.json) {
      console.log(JSON.stringify({ port, range: [start, end] }, null, 2));
      return;
    }
    console.log(port);
  });

program
  .command('portmasters')
  .description('list PORTMASTER.md files')
  .option('--json', 'output as JSON')
  .action((opts) => {
    const files = getPortmasterFiles();
    if (opts.json) {
      console.log(JSON.stringify(files, null, 2));
      return;
    }
    if (files.length === 0) {
      console.log('No PORTMASTER.md files found.');
      return;
    }
    for (const f of files) {
      console.log(f);
    }
  });

program.parse();
